src/ray/raylet/scheduling/cluster_task_manager_test.cc

```diff
@@ -1557,6 +1615,440 @@ TEST_F(ClusterTaskManagerTest, PopWorkerExactlyOnce) {
+      },
+      /*max_pinned_task_arguments_bytes=*/1000);
+
+  RayTask task = CreateTask({}, /*num_args=*/0, /*args=*/{});
+  RayTask task2 = CreateTask({}, /*num_args=*/0, /*args=*/{});
+  RayTask task3 = CreateTask({}, /*num_args=*/0, /*args=*/{});
+  auto runtime_env_hash = task.GetTaskSpecification().GetRuntimeEnvHash();
+  std::vector<std::shared_ptr<MockWorker>> workers;
+  for (int i = 0; i < 3; i++) {
+    std::shared_ptr<MockWorker> worker =
+        std::make_shared<MockWorker>(WorkerID::FromRandom(), 1234, runtime_env_hash);
+    pool_.PushWorker(std::static_pointer_cast<WorkerInterface>(worker));
+    pool_.TriggerCallbacks();
+    workers.push_back(worker);
+  }
+  rpc::RequestWorkerLeaseReply reply;
+  int num_callbacks = 0;
+  auto callback = [&num_callbacks](Status, std::function<void()>, std::function<void()>) {
+    num_callbacks++;
+  };
+  task_manager_.QueueAndScheduleTask(task, &reply, callback);
+  task_manager_.QueueAndScheduleTask(task2, &reply, callback);
+  task_manager_.QueueAndScheduleTask(task3, &reply, callback);
+  pool_.TriggerCallbacks();
+
+  // We shouldn't cap anything for zero cpu tasks (and shouldn't crash before
+  // this point).
+  ASSERT_EQ(num_callbacks, 3);
+
+  for (auto &worker : workers) {
+    RayTask buf;
+    task_manager_.TaskFinished(worker, &buf);
+  }
+
+  AssertNoLeaks();
+}
+
+/// Test that we exponentially increase the amount of time it takes to increase
+/// the dispatch cap for a scheduling class.
+TEST_F(ClusterTaskManagerTest, SchedulingClassCapIncrease) {
+  auto get_unblocked_worker = [](std::vector<std::shared_ptr<MockWorker>> &workers)
+      -> std::shared_ptr<MockWorker> {
+    for (auto &worker : workers) {
+      if (worker->GetAllocatedInstances() != nullptr && !worker->IsBlocked()) {
+        return worker;
+      }
+    }
+    return nullptr;
+  };
+
+  int64_t UNIT = RayConfig::instance().worker_cap_initial_backoff_delay_ms();
+  std::vector<RayTask> tasks;
+  for (int i = 0; i < 3; i++) {
+    RayTask task = CreateTask({{ray::kCPU_ResourceLabel, 8}},
+                              /*num_args=*/0, /*args=*/{});
+    tasks.emplace_back(task);
+  }
+
+  rpc::RequestWorkerLeaseReply reply;
+  int num_callbacks = 0;
+  auto callback = [&num_callbacks](Status, std::function<void()>, std::function<void()>) {
+    num_callbacks++;
+  };
+  for (const auto &task : tasks) {
+    task_manager_.QueueAndScheduleTask(task, &reply, callback);
+  }
+
+  auto runtime_env_hash = tasks[0].GetTaskSpecification().GetRuntimeEnvHash();
+  std::vector<std::shared_ptr<MockWorker>> workers;
+  for (int i = 0; i < 3; i++) {
+    std::shared_ptr<MockWorker> worker =
+        std::make_shared<MockWorker>(WorkerID::FromRandom(), 1234, runtime_env_hash);
+    pool_.PushWorker(std::static_pointer_cast<WorkerInterface>(worker));
+    pool_.TriggerCallbacks();
+    workers.push_back(worker);
+  }
+  task_manager_.ScheduleAndDispatchTasks();
+
+  ASSERT_EQ(num_callbacks, 1);
+
+  current_time_ms_ += UNIT;
```

Comment:
Just skimming this, it does look like the case I mentioned above is not covered here. So we should  also check that calling ScheduleAndDispatchTasks while the timer is set has no effect on the timer.

Reply:
Doesn't the block starting on line 1831 do that? at that point the time is  200ms, and the wait is set to 300ms. On 1840, it goes up to 300ms and then the task runs.